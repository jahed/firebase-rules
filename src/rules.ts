import { RuleExpression } from './types'

/**
 * Returns a strictly `true` expression.
 *
 * `read(allowAll)` is the same as `{ ".read": false }`
 */
const allowAll: RuleExpression<boolean, boolean> = () => true

/**
 * Returns a strictly `false` expression.
 *
 * `read(denyAll)` is the same as `{ ".read": false }`
 */
const denyAll: RuleExpression<boolean, boolean> = () => false

export { allowAll, denyAll }
