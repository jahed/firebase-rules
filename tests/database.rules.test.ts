import { rules } from './database.rules'

test('json rules are up-to-date', () => {
  expect(JSON.stringify(rules(), null, 2)).toMatchSnapshot()
})
