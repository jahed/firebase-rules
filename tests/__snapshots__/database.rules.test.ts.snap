// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json rules are up-to-date 1`] = `
"{
  \\"rules\\": {
    \\"app\\": {
      \\"update\\": {
        \\"version\\": {
          \\".validate\\": \\"newData.isString()\\"
        },
        \\"force\\": {
          \\".validate\\": \\"newData.isBoolean()\\"
        },
        \\"timestamp\\": {
          \\".validate\\": \\"newData.isNumber()\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        }
      },
      \\"maintenance\\": {
        \\"message\\": {
          \\".validate\\": \\"newData.isString()\\"
        },
        \\"enabled\\": {
          \\".validate\\": \\"newData.isBoolean()\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        }
      },
      \\"postDelay\\": {
        \\".validate\\": \\"newData.isNumber()\\"
      },
      \\"$other\\": {
        \\".validate\\": false
      },
      \\".read\\": true,
      \\".write\\": false
    },
    \\"bookmarks\\": {
      \\"$userId\\": {
        \\"$pathname\\": {
          \\"name\\": {
            \\".validate\\": \\"(newData.isString() && ((newData.val().length > 0) && (newData.val().length < 120)))\\"
          },
          \\"pathname\\": {
            \\".validate\\": \\"(newData.isString() && ((newData.val().length > 0) && (newData.val().length < 1000)))\\"
          },
          \\"created_at\\": {
            \\".validate\\": \\"(newData.isNumber() && (!data.exists() || (data.isNumber() && (data.val() === newData.val()))))\\"
          },
          \\"$other\\": {
            \\".validate\\": false
          },
          \\".validate\\": \\"newData.hasChildren([\\\\\\"name\\\\\\",\\\\\\"pathname\\\\\\",\\\\\\"created_at\\\\\\"])\\"
        },
        \\".read\\": \\"($userId === auth.uid)\\",
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"completedNodes\\": {
      \\"$userId\\": {
        \\"$gameId\\": {
          \\"$nodeId\\": {
            \\".write\\": \\"($userId === auth.uid)\\",
            \\".validate\\": \\"(newData.isNumber() && (!data.exists() || (data.isNumber() && (data.val() === newData.val()))))\\"
          },
          \\".read\\": true,
          \\".validate\\": \\"root.child(\\\\\\"games/\\\\\\" + $gameId).exists()\\"
        },
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"gameUsers\\": {
      \\"$gameId\\": {
        \\"$userId\\": {
          \\".write\\": \\"($userId === auth.uid)\\",
          \\".validate\\": \\"(newData.isNumber() && (!data.exists() || (data.isNumber() && (newData.val() > data.val()))))\\"
        },
        \\".validate\\": \\"root.child(\\\\\\"games/\\\\\\" + $gameId).exists()\\",
        \\".read\\": true,
        \\".indexOn\\": [
          \\".value\\"
        ]
      }
    },
    \\"notes\\": {
      \\"$userId\\": {
        \\"text\\": {
          \\".validate\\": \\"(newData.isString() && (newData.val().length < 1001))\\"
        },
        \\"updated_at\\": {
          \\".validate\\": \\"newData.isNumber()\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        },
        \\".read\\": \\"($userId === auth.uid)\\",
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"(root.child(\\\\\\"users/\\\\\\" + $userId).exists() && newData.hasChildren([\\\\\\"text\\\\\\",\\\\\\"updated_at\\\\\\"]))\\"
      }
    },
    \\"preferences\\": {
      \\"$userId\\": {
        \\"audio\\": {
          \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
        },
        \\"locale\\": {
          \\".validate\\": \\"(newData.isString() && (newData.val().length === 2))\\"
        },
        \\"quality_id\\": {
          \\".validate\\": \\"(newData.isString() && (newData.val().length < 10))\\"
        },
        \\"editor\\": {
          \\".validate\\": \\"newData.isBoolean()\\"
        },
        \\"hide_presence\\": {
          \\".validate\\": \\"newData.isBoolean()\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        },
        \\".read\\": \\"($userId === auth.uid)\\",
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"probeGuides\\": {
      \\"$userId\\": {
        \\"$probe_guide\\": {
          \\"name\\": {
            \\".validate\\": \\"(newData.isString() && ((newData.val().length > 0) && (newData.val().length < 50)))\\"
          },
          \\"siteProbeMap\\": {
            \\"$siteNumber\\": {
              \\".validate\\": \\"($siteNumber.matches(/^[12345]\\\\\\\\\\\\\\\\d{2}$/) && newData.isNumber())\\"
            }
          },
          \\"$other\\": {
            \\".validate\\": false
          },
          \\".validate\\": \\"newData.hasChildren([\\\\\\"name\\\\\\",\\\\\\"siteProbeMap\\\\\\"])\\"
        },
        \\".read\\": true,
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"userGames\\": {
      \\"$userId\\": {
        \\"$gameId\\": {
          \\"playState\\": {
            \\"current\\": {
              \\".validate\\": \\"(newData.isString() && (newData.val().matches(/^[a-z-]+$/) && (newData.val().length < 32)))\\"
            },
            \\"body\\": {
              \\".validate\\": \\"(newData.isString() && ((newData.val().length < 201) && !newData.val().contains(\\\\\\"\\\\\\\\n\\\\\\")))\\"
            },
            \\"updatedAt\\": {
              \\".validate\\": \\"newData.isNumber()\\"
            },
            \\"$other\\": {
              \\".validate\\": false
            }
          },
          \\"$other\\": {
            \\".validate\\": false
          },
          \\".write\\": \\"($userId === auth.uid)\\",
          \\".validate\\": \\"root.child(\\\\\\"games/\\\\\\" + $gameId).exists()\\"
        },
        \\".read\\": true,
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"users\\": {
      \\"$userId\\": {
        \\"name\\": {
          \\".validate\\": \\"(newData.isString() && ((newData.val().length > 0) && (newData.val().length < 24)))\\"
        },
        \\"created_at\\": {
          \\".validate\\": \\"(newData.isNumber() && (!data.exists() || (data.isNumber() && (data.val() === newData.val()))))\\"
        },
        \\"avatar_id\\": {
          \\".validate\\": \\"(newData.isString() && (newData.val().matches(/^[0-9a-z-]+$/) && (newData.val().length < 32)))\\"
        },
        \\"hue\\": {
          \\".validate\\": \\"(newData.isNumber() && ((newData.val() > -1) && (newData.val() < 360)))\\"
        },
        \\"about\\": {
          \\".validate\\": \\"(!newData.exists() || (newData.isString() && (newData.val().length < 140)))\\"
        },
        \\"location\\": {
          \\".validate\\": \\"(!newData.exists() || (newData.isString() && (newData.val().length < 32)))\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        },
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"newData.hasChildren([\\\\\\"name\\\\\\",\\\\\\"created_at\\\\\\"])\\"
      },
      \\".read\\": true
    },
    \\"games\\": {
      \\"$gameId\\": {
        \\".read\\": false,
        \\".write\\": false,
        \\".validate\\": \\"(newData.isBoolean() && (newData.val() === true))\\"
      }
    },
    \\"gamePosts\\": {
      \\"$gameId\\": {
        \\"$postId\\": {
          \\".write\\": \\"((root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").val() === auth.uid)) && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/gameId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/gameId\\\\\\").val() === $gameId)))\\",
          \\".validate\\": \\"(!data.exists() && (newData.isNumber() && (newData.val() === now)))\\"
        },
        \\".indexOn\\": [
          \\".value\\"
        ],
        \\".read\\": true,
        \\".validate\\": \\"root.child(\\\\\\"games/\\\\\\" + $gameId).exists()\\"
      }
    },
    \\"userPosts\\": {
      \\"$userId\\": {
        \\"$postId\\": {
          \\".write\\": \\"(($userId === auth.uid) && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").val() === auth.uid)))\\",
          \\".validate\\": \\"(!data.exists() && (newData.isNumber() && (newData.val() === now)))\\"
        },
        \\".indexOn\\": [
          \\".value\\"
        ],
        \\".read\\": true,
        \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
      }
    },
    \\"gameUserPosts\\": {
      \\"$gameId\\": {
        \\"$userId\\": {
          \\"$postId\\": {
            \\".write\\": \\"((root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/userId\\\\\\").val() === auth.uid)) && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/gameId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $postId + \\\\\\"/gameId\\\\\\").val() === $gameId)))\\",
            \\".validate\\": \\"(!data.exists() && (newData.isNumber() && (newData.val() === now)))\\"
          },
          \\".indexOn\\": [
            \\".value\\"
          ],
          \\".read\\": true,
          \\".validate\\": \\"root.child(\\\\\\"users/\\\\\\" + $userId).exists()\\"
        },
        \\".validate\\": \\"root.child(\\\\\\"games/\\\\\\" + $gameId).exists()\\"
      }
    },
    \\"postReplies\\": {
      \\"$postId\\": {
        \\"$replyId\\": {
          \\".write\\": \\"((root.child(\\\\\\"posts/\\\\\\" + $replyId + \\\\\\"/userId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $replyId + \\\\\\"/userId\\\\\\").val() === auth.uid)) && (root.child(\\\\\\"posts/\\\\\\" + $replyId + \\\\\\"/parentId\\\\\\").isString() && (root.child(\\\\\\"posts/\\\\\\" + $replyId + \\\\\\"/parentId\\\\\\").val() === $postId)))\\",
          \\".validate\\": \\"(!data.exists() && (newData.isNumber() && (newData.val() === now)))\\"
        },
        \\".indexOn\\": [
          \\".value\\"
        ],
        \\".read\\": true,
        \\".validate\\": \\"root.child(\\\\\\"posts/\\\\\\" + $postId).exists()\\"
      }
    },
    \\"posts\\": {
      \\"$postId\\": {
        \\"createdAt\\": {
          \\".validate\\": \\"((data.exists() && (newData.isNumber() && (data.isNumber() && (newData.val() === data.val())))) || (!data.exists() && (newData.isNumber() && (newData.val() === now))))\\"
        },
        \\"updatedAt\\": {
          \\".validate\\": \\"(newData.isNumber() && (newData.val() === now))\\"
        },
        \\"userId\\": {
          \\".validate\\": \\"((data.exists() && (newData.isString() && (data.isString() && (newData.val() === data.val())))) || (!data.exists() && (newData.isString() && root.child(\\\\\\"users/\\\\\\" + newData.val()).exists())))\\"
        },
        \\"gameId\\": {
          \\".validate\\": \\"((data.exists() && (newData.isString() && (data.isString() && (newData.val() === data.val())))) || (!data.exists() && (newData.isString() && root.child(\\\\\\"games/\\\\\\" + newData.val()).exists())))\\"
        },
        \\"parentId\\": {
          \\".validate\\": \\"((root.child(\\\\\\"posts/\\\\\\" + $postId).exists() && (newData.isString() && (data.isString() && (newData.val() === data.val())))) || (!root.child(\\\\\\"posts/\\\\\\" + $postId).exists() && (!newData.exists() || (newData.isString() && root.child(\\\\\\"posts/\\\\\\" + newData.val()).exists()))))\\"
        },
        \\"body\\": {
          \\".validate\\": \\"(newData.isString() && ((newData.val().length > 0) && (newData.val().length < 1024)))\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        },
        \\".read\\": true,
        \\".write\\": \\"((newData.child(\\\\\\"userId\\\\\\").isString() && (newData.child(\\\\\\"userId\\\\\\").val() === auth.uid)) && (root.child(\\\\\\"postRequests/\\\\\\" + auth.uid + \\\\\\"/postId\\\\\\").isString() && (root.child(\\\\\\"postRequests/\\\\\\" + auth.uid + \\\\\\"/postId\\\\\\").val() === $postId)))\\",
        \\".validate\\": \\"(!data.exists() && newData.hasChildren([\\\\\\"createdAt\\\\\\",\\\\\\"updatedAt\\\\\\",\\\\\\"userId\\\\\\",\\\\\\"gameId\\\\\\",\\\\\\"body\\\\\\"]))\\"
      }
    },
    \\"postRequests\\": {
      \\"$userId\\": {
        \\"createdAt\\": {
          \\".validate\\": \\"(newData.isNumber() && (newData.val() === now))\\"
        },
        \\"postId\\": {
          \\".validate\\": \\"!(newData.isString() && root.child(\\\\\\"posts/\\\\\\" + newData.val()).exists())\\"
        },
        \\"$other\\": {
          \\".validate\\": false
        },
        \\".read\\": \\"($userId === auth.uid)\\",
        \\".write\\": \\"($userId === auth.uid)\\",
        \\".validate\\": \\"(newData.hasChildren([\\\\\\"createdAt\\\\\\"]) && (!data.exists() || (data.child(\\\\\\"createdAt\\\\\\").isNumber() && (root.child(\\\\\\"app/postDelay\\\\\\").isNumber() && (data.child(\\\\\\"createdAt\\\\\\").val() < (now - root.child(\\\\\\"app/postDelay\\\\\\").val()))))))\\"
      }
    },
    \\"authTest\\": {
      \\"uid\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.uid))\\"
      },
      \\"provider\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.provider))\\"
      },
      \\"$other\\": {
        \\".validate\\": false
      }
    },
    \\"authTokenTest\\": {
      \\"email\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.email))\\"
      },
      \\"email_verified\\": {
        \\".validate\\": \\"(newData.isBoolean() && (newData.val() === auth.token.email_verified))\\"
      },
      \\"phone_number\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.phone_number))\\"
      },
      \\"name\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.name))\\"
      },
      \\"sub\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.sub))\\"
      },
      \\"iss\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.iss))\\"
      },
      \\"aud\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.aud))\\"
      },
      \\"auth_time\\": {
        \\".validate\\": \\"(newData.isNumber() && (newData.val() === auth.token.auth_time))\\"
      },
      \\"iat\\": {
        \\".validate\\": \\"(newData.isNumber() && (newData.val() === auth.token.iat))\\"
      },
      \\"exp\\": {
        \\".validate\\": \\"(newData.isNumber() && (newData.val() === auth.token.exp))\\"
      },
      \\"$other\\": {
        \\".validate\\": false
      }
    },
    \\"authTokenFirebaseTest\\": {
      \\"sign_in_provider\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase.sign_in_provider))\\"
      },
      \\"$other\\": {
        \\".validate\\": false
      }
    },
    \\"authTokenFirebaseIdentitiesTest\\": {
      \\"email\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"email\\\\\\"][0]))\\"
      },
      \\"phone\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"phone\\\\\\"][1]))\\"
      },
      \\"google_com\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"google.com\\\\\\"][2]))\\"
      },
      \\"facebook_com\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"facebook.com\\\\\\"][3]))\\"
      },
      \\"github_com\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"github.com\\\\\\"][4]))\\"
      },
      \\"twitter_com\\": {
        \\".validate\\": \\"(newData.isString() && (newData.val() === auth.token.firebase[\\\\\\"twitter.com\\\\\\"][5]))\\"
      },
      \\"$other\\": {
        \\".validate\\": false
      }
    },
    \\"$other\\": {
      \\".validate\\": false
    }
  }
}"
`;
